zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+        let address = TransparentAddress::decode(params, &addr_str)?;
+        result.insert(address, Some(metadata(address_index)));
+    }
+
+    if for_detection {
+        if let Some(first) = first_unused_index {
+            let ephemeral_ivk = get_ephemeral_ivk(conn, params, account_id)?;
+
+            for raw_index in range_after(first, GAP_LIMIT) {
+                let address_index = NonHardenedChildIndex::from_index(raw_index).unwrap();
+                let address = ephemeral_ivk.derive_ephemeral_address(address_index)?;
+                result.insert(address, Some(metadata(address_index)));
+            }
+        }
+    }
+    Ok(result)
+}
+
+/// Returns a vector with the next `n` previously unreserved ephemeral addresses for
+/// the given account.
+///
+/// Precondition: `n < 0x80000000`
+///
+/// # Errors
+///
+/// * `SqliteClientError::AccountUnknown`, if there is no account with the given id.
+/// * `SqliteClientError::UnknownZip32Derivation`, if the account is imported and
+///   it is not possible to derive new addresses for it.
+/// * `SqliteClientError::ReachedGapLimit`, if it is not possible to reserve `n` addresses
+///   within the gap limit after the last address in this account that is known to have an
+///   output in a mined transaction.
+/// * `SqliteClientError::AddressGeneration(AddressGenerationError::DiversifierSpaceExhausted)`,
+///   if the limit on transparent address indices has been reached.
+pub(crate) fn reserve_next_n_ephemeral_addresses<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    account_id: AccountId,
+    n: u32,
+) -> Result<Vec<(TransparentAddress, TransparentAddressMetadata)>, SqliteClientError> {
+    if n == 0 {
+        return Ok(vec![]);
+    }
+    assert!(n > 0);
+    let n = i32::try_from(n).expect("precondition violated");
+
+    let ephemeral_ivk = get_ephemeral_ivk(wdb.conn.0, &wdb.params, account_id)?;
+    let last_reserved_index = last_reserved_index(wdb.conn.0, account_id)?;
+    let last_safe_index = last_safe_index(wdb.conn.0, account_id)?;
+    let allocation = range_after(last_reserved_index, n);
+
+    if allocation.clone().count() < n.try_into().unwrap() {
+        return Err(SqliteClientError::AddressGeneration(
+            AddressGenerationError::DiversifierSpaceExhausted,
+        ));
+    }
+    if *allocation.end() > last_safe_index {
+        let unsafe_index = max(*allocation.start(), last_safe_index.saturating_add(1));
+        return Err(SqliteClientError::ReachedGapLimit(account_id, unsafe_index));
+    }
+
+    // used_in_tx and mined_in_tx are initially NULL
+    let mut stmt_insert_ephemeral_address = wdb.conn.0.prepare_cached(
+        "INSERT INTO ephemeral_addresses (account_id, address_index, address)
+         VALUES (:account_id, :address_index, :address)",
+    )?;
+
+    allocation
+        .map(|raw_index| {
+            let address_index = NonHardenedChildIndex::from_index(raw_index).unwrap();
```

Comment:
Per my previous comment, this is the only place we insert into `ephemeral_addresses` currently, and we only insert precisely the range that was reserved. By construction here `allocation` connects to the immediate earlier range without leaving a gap, and thus currently we don't create internal gaps.